### DISCUSSION TITLE
NameError, what is the logic behind checking against protected namespaces?

### DISCUSSION THREAD

**Discussion by nss350** [pydantic/pydantic#7763] [Question]
Hi,
I've was writing a class using pydantic which has an attribute named "params_". When I try and build documentation for the class using autodoc pydantic, the following error gets raised in pydantic._internal._fields.collect_model_fields.

NameError: Field "model_fields" conflicts with member {'params_': FieldInfo(annotation=ndarray, required=False)} of protected namespace "model_".

From reading about protected namespaces, I don't get the impression that the attribute name "params_" should flag this error, but putting this here as it may not be a bug.

Any suggestions appreciated.

**Comment by guibos** [comment]
@nss350 My case: it is produced when I try to inherit from a parent class. Parent class does not raise the problem.

**Comment by MSSputnik** [comment]
Today I debugged the error I ignored for a while. 
It looks like that due to the inheritance the field `model_field` get set in the parent class. Maybe it is set when sphinx or autodoc processes the parent class. I did not find out.
My work around was to import the problematic module in conftest.py, not the class.
Sphinx is doing "importlib.import_module()" and with this, the module is already loaded and not imported again, not causing the problematic check from pydantic.

**Reply by samanthavbarron** [reply]
Could you please elaborate on the workaround that you found? I am not sure I understand which module you're referring to.

**Reply by swinkels** [reply]
I faced the same issue. autodoc raised the NameError when it tried to import a module (I wrote) that referenced a class that 1) inherited from the Pydantic BaseModel and 2) used generics. I worked around the NameError by explicitly importing that module in `conf.py`, so in the Sphinx configuration file.

I looked into the underlying cause and got to the same point as mentioned [in the comment above](https://github.com/pydantic/pydantic/discussions/7763#discussioncomment-8417097). By then you're deep into Pydantic internals and in the end I just settled for the workaround.

**Reply by FredStober** [reply]
I have the same issue - but I'm not sure where to open a bug report about this - sphinx or pydantic....
It started happening when changing to pydantic 2 ... But it's happening in so many places in my code base that it's just not feasible to import everything in conf.py.

**Reply by Yalchin403** [reply]
Thanks @swinkels. I think importing the inherited schemas are the easiest way to get around this issue. But i think this is something to be resolved on the pydantic end.

**Reply by swinkels** [reply]
Thank you for the shoutout, but the workaround is not my idea. Credit should go to @MSSputnik for that one.

**Comment by SallyDa** [comment]
Another solution is to add `model_config = ConfigDict(protected_namespaces=())` inside the problematic pydantic class.

**Comment by gVallverdu** [comment]
I got the same error.

```WARNING: autodoc: failed to import module 'sample' from module 'controls'; the following exception was raised:
Traceback (most recent call last):
  File "/Users/gvallver/opt/anaconda3/envs/c2mc10/lib/python3.10/site-packages/sphinx/ext/autodoc/importer.py", line 143, in import_module
    return importlib.import_module(modname)
  File "/Users/gvallver/opt/anaconda3/envs/c2mc10/lib/python3.10/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/Users/gvallver/git/pyc2mc/src/pyc2mc/controls/sample.py", line 20, in <module>
    class SampleImportControls(ProcessingControls):
  File "/Users/gvallver/opt/anaconda3/envs/c2mc10/lib/python3.10/site-packages/pydantic/_internal/_model_construction.py", line 183, in __new__
    set_model_fields(cls, bases, config_wrapper, types_namespace)
  File "/Users/gvallver/opt/anaconda3/envs/c2mc10/lib/python3.10/site-packages/pydantic/_internal/_model_construction.py", line 418, in set_model_fields
    fields, class_vars = collect_model_fields(cls, bases, config_wrapper, types_namespace, typevars_map=typevars_map)
  File "/Users/gvallver/opt/anaconda3/envs/c2mc10/lib/python3.10/site-packages/pydantic/_internal/_fields.py", line 120, in collect_model_fields
    raise NameError(
NameError: Field "model_fields" conflicts with member {} of protected namespace "model_".
```

Running `make html` a second times, makes the error to disappear. This is not a solution, but the docs is complete.
I got the same error with python 3.11.

```
nbsphinx==0.9.4
pydantic==2.4.2
pydantic_core==2.10.1
Sphinx==7.3.7
sphinxcontrib-applehelp==1.0.8
sphinxcontrib-devhelp==1.0.6
sphinxcontrib-htmlhelp==2.0.5
sphinxcontrib-jquery==4.1
sphinxcontrib-jsmath==1.0.1
sphinxcontrib-qthelp==1.0.7
sphinxcontrib-serializinghtml==1.1.10
```
